### PR TITLE
re-implement the task retry state using xtriggers

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Remove the following task states:
  * `TASK_STATUS_RETRYING`
  * `TASK_STATUS_SUBMIT_RETRYING`
* Re-implement retry logic using xtriggers.
  * Create an xtirigger called `cylc[_submission]_retry_<task>.<cycle>`
    * Use the `cylc` namespace for special triggers.
  * Create only one xtrigger per task per retry type.
* Reset "retrying" tasks to waiting.

Caveats:
* Can no longer select tasks by retrying states on the CLI (e.g. `<cyclepoint>/*:retrying`)
* Retrying tasks are now waiting so have a different position in `TASK_STATUS_ORDERED`
  * TODO need to check this doesn't break polling logic

These changes partially address https://github.com/cylc/cylc-admin/pull/47

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] I have opened a documentation Issue at cylc/cylc-doc#162.

### CODE REVIEW

**Comment by hjoliver** [comment]
That was quick!

**Comment by hjoliver** [comment]
I'm part way through reviewing this.  I think it's all good, but there's a few caveats to think through, as you've pointed out.

File: cylc/flow/task_events_mgr.py [hunk 1]
```diff
@@ -675,6 +686,67 @@ def _job_logs_retrieval_callback(self, proc_ctx, schd_ctx):
...
+            xtrig = SubFuncContext(
+                label,
+                'wall_clock',
+                [],
+                kwargs
+            )
+            self.xtrigger_mgr.add_trig(
+                label,
+                xtrig,
+                os.getenv("CYLC_SUITE_RUN_DIR")
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_events_mgr.py]: I suppose so. The server should have exported that in its own environment. It ain't pretty though - can we have a sort of "workflow context" module to import where needed?

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply cylc/flow/task_events_mgr.py]: I think `cylc.flow.suite_files` should handle this (one day), see subnotes on #3345.

File: cylc/flow/task_events_mgr.py [hunk 2]
```diff
@@ -477,8 +489,26 @@ def _process_message_check(
...
+                    (
+                        # task has automatically execn-retried at least once
+                        TimerFlags.EXECUTION_RETRY in itask.try_timers
+                        and itask.try_timers[
+                            TimerFlags.EXECUTION_RETRY].num > 0
+                    )
+                )
+
         ):
             # Ignore polled messages if task is already in retrying statuses
```
Comment (hjoliver) [inline cylc/flow/task_events_mgr.py]:
```suggestion
            # Ignore polled messages if task has a retry lined up
```

File: cylc/flow/task_events_mgr.py [hunk 3]
```diff
@@ -477,8 +489,26 @@ def _process_message_check(
...
             return False
-        if itask.state.status in (
-            TASK_STATUS_SUBMIT_RETRYING, TASK_STATUS_RETRYING
+
+        if (
+                itask.state(TASK_STATUS_WAITING)
+                and
+                (
+                    (
+                        # task has automatically submit-retried at least once
```
Comment (hjoliver) [inline cylc/flow/task_events_mgr.py]:
```suggestion
                        # task has a submit-retry lined up
```
? ("has submit-retried at least one" suggests to me it has already RE-submitted at least once")

File: cylc/flow/task_events_mgr.py [hunk 4]
```diff
@@ -477,8 +489,26 @@ def _process_message_check(
...
+                        TimerFlags.SUBMISSION_RETRY in itask.try_timers
+                        and itask.try_timers[
+                            TimerFlags.SUBMISSION_RETRY].num > 0
+                    )
+                    or
+                    (
+                        # task has automatically execn-retried at least once
```
Comment (hjoliver) [inline cylc/flow/task_events_mgr.py]:
```suggestion
                        # task has an execution-retry lined up
```
? (ditto my submit-retry comment)

**Comment by hjoliver** [comment]
@oliver-sanders - did you have a reason to think this might break polling logic?  (If you remember - sorry it took so long!!)  It looks to me as if there is a 1:1 mapping between the old retry task status and the new "waiting status, with an xtrigger" ... so it shouldn't make any difference to polling. But maybe I missed something.

**Comment by hjoliver** [comment]
(Ah, it was to do with position in TASK_STATUS_ORDERED?)

**Comment by hjoliver** [comment]
Tests all passed, after kicking the flaky batch.

**Comment by oliver-sanders** [comment]
Putting this up against 8.0a4 in the misguided hope that this PR won't reach its first birthday.

**Comment by hjoliver** [comment]
I'm taking this PR over so that @oliver-sanders can concentrate on other things.  Plus, many of the conflicts result from my big SoD PR some months back. Oliver says he took a look at my conflict resolution and it looked "legit", so I'll see if can get the tests passing, then enlist another reviewer.

First: this branch backed up as-is to https://github.com/hjoliver/cylc-flow/tree/oliver-retry-orig

**Comment by kinow** [comment]
Happy to help with reviewing later, if necessary. The task states are used extensively in the UI, so that may help me if I need to update the UI later too.

**Comment by kinow** [comment]
bash failures are related to docker and memory, see #3856 for fix

```
2020-10-07T04:25:41.9392152Z    write (2, "virtual memory exhausted\n", 25);
```

**Comment by hjoliver** [comment]
All functional and unit tests passing now.

The bash test `tests/functional/broadcast/00-simple.t` fails intermittently on other branches too.

**Comment by hjoliver** [comment]
(The last 3 commits just modify comments, via the GH UI "commit suggestion" button).

File: cylc/flow/task_events_mgr.py [hunk 5]
```diff
@@ -677,6 +706,51 @@ def _job_logs_retrieval_callback(self, proc_ctx, schd_ctx):
...
+            itask.identity
+        ))
+        kwargs = {
+            'absolute_as_seconds': wallclock_time
+        }
+
+        # if this isn't the first retry the xtrigger will already exist
+        if label in itask.state.xtriggers:
+            # retry xtrigger already exists from a previous retry, modify it
+            xtrig = itask.state.xtriggers[label]
```
Comment (kinow) [inline cylc/flow/task_events_mgr.py:735]: `xtrig` never used on this side of the `if` statement?

**Comment by hjoliver** [comment]
(All tests passing :grin: )

File: cylc/flow/xtrigger_mgr.py [hunk 6]
```diff
@@ -86,32 +86,17 @@ class XtriggerManager:
...
 
-    """
+    Args:
+        suite (str): suite name
+        user (str): suite owner
+        broadcast_mgr (BroadcastMgr): the Broadcast Manager
+        proc_pool (SubProcPool): pool of Subprocesses
+        suite_run_dir (str): suite run directory
+        suite_share_dir (str): suite share directory
+        suite_work_dir (str): suite work directory
```
Comment (kinow) [inline cylc/flow/xtrigger_mgr.py:96]: No `suite_work_dir` in constructor? I think this one can be removed.

File: cylc/flow/rundb.py [hunk 7]
```diff
@@ -916,6 +917,58 @@ def remove_columns(self, table, to_drop):
...
+                '''
+            ))
+            if tasks:
+                LOG.info(f'Upgrade retrying tasks in table {table}')
+            conn.executemany(
+                rf'''
+                    UPDATE
+                        {table}
+                    SET
+                        status='{TASK_STATUS_WAITING}'
```
Comment (kinow) [inline cylc/flow/rundb.py:961]: `self.TASK_STATUS_WAITING` ? The GitHub UI is not highlighting this part of the text (I see the replaced-variables in black, while the rest of string-text is in blue). Not sure if we have a test for this (last I can remember some SQL code being tested I think it was in a change by @wxtim )

**Review by kinow — COMMENTED** [review]
A few typos and pydocs issues. No issues with rest of the code. I did find a few interesting things in other parts of the code, but they were not in this pull request (which means I've got more post-its on my screen now to investigate later :grimacing: )

@hjoliver I tested it with Cylc UIServer, and found no exceptions. With Cylc UI, however, the `complex` workflow logged that error due to missing parents.

image

image

I tested `master` and the errors didn't occur. But given how little we know about that error, it could be that the same happens on `master`, but intermittently. I have no idea if `complex` has anything special about it with regards to the change of retry state in this PR :+1: 

Thanks!
B

File: cylc/flow/rundb.py (hunk 7 above)
Comment (oliver-sanders) [reply cylc/flow/rundb.py:961]: There was a test for this, I think it got lost in the merge, have pushed a commit to re-instate it.

This PR was written back when we were still aiming for `cylc restart` compatibility with Cylc7. We've given up on that one now (good idea due to major changes) so this upgrade code can probably be dropped.

**Comment by hjoliver** [comment]
>  I have no idea if complex has anything special about it with regards to the change of retry state in this PR

I just checked, it doesn't look like any tasks in the complex suite will retry so it must be a coincidence.

File: cylc/flow/task_events_mgr.py [hunk 8]
```diff
@@ -694,18 +768,22 @@ def _process_message_failed(self, itask, event_time, message):
...
                 "[%s] -job(%02d) %s", itask, itask.submit_num, "failed")
             no_retries = True
-        elif itask.state.reset(TASK_STATUS_RETRYING):
-            delay_msg = "retrying in %s" % (
-                itask.try_timers[TASK_STATUS_RETRYING].delay_timeout_as_str())
+        else:
+            # There is an execution retry lined up.
+            timer = itask.try_timers[TimerFlags.EXECUTION_RETRY]
+            self._retry_task(itask, timer.timeout)
+            delay_msg = f"retrying in {timer.delay_timeout_as_str()}"
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_events_mgr.py:785]: Really both are just a "retry", it's just that one is motivated by a submit failure, and one by an execution failure. Historically we've only singled out submit-retry like this, which is probably OK.

File: cylc/flow/rundb.py (hunk 7 above)
Comment (kinow) [reply cylc/flow/rundb.py:961]: @hjoliver I think this is the last item pending before I hit approve. Should we just drop the database migration here?

Comment (hjoliver) [reply cylc/flow/rundb.py:961]: Yeah just looking at this, but I'm slightly confused. Oliver re-instated the DB test, but doesn't that test the functionality he said we can drop now??

Comment (hjoliver) [reply cylc/flow/rundb.py:961]: It seems to me @oliver-sanders reinstated the DB test before realising we can drop the upgrade code.  So I'll remove it again (and fix the conflicts).

Comment (hjoliver) [reply cylc/flow/rundb.py:961]: Cancel that, I've fixed the conflicts but will post a note to Element about upgrade from Cylc 7...

**Review by kinow — APPROVED** [review]
Had a look at the files that GH UI showed as changed since my last review (thanks GH for such handy functionality). Looking good to me, +1

**Comment by hjoliver** [comment]
@kinow - the same flaky bask test failed again. I think you can approve and merge this if the functional tests all pass.

**Comment by kinow** [comment]
(@hjoliver I'm guessing CI is still a bit unstable, but approved since the conflict changed files were only 3, from a PR that I reviewed yesterday; no issues that I could see)

**Comment by kinow** [comment]
> @kinow - the same flaky bask test failed again. I think you can approve and merge this if the functional tests all pass.

Missed this comment by 23 seconds, while I was typing. But I managed to approve before your comment :smile:

**Review by hjoliver — APPROVED** [review]

**Comment by hjoliver** [comment]
(Two approvals :grin: and the bash test fail is a known flaky ... merging)

**Comment by oliver-sanders** [comment]
Wow, did that really happen.

**Pull request merged**